File: comparepythoncfortran/jacobiRelax/jacobirelax.f90
program main
  integer, parameter :: ssize = 2048
  real, dimension(ssize,ssize)::u
  real, dimension(ssize,ssize)::f
  real, dimension(ssize,ssize)::u_new
  real, dimension(ssize,ssize)::temp

  call zero(u)
  call zero(f)
  call zero(u_new)
  call random(u)
  do k=1, 10
    do i=2, 2047
      do j=2, 2047
        u_new(i,j) = 0.25*(h2*f(i,j) + u(i-1,j) + u(i+1,j) + u(i,j-1) + u(i,j+1))
      enddo
    enddo
    write (*,*) "Iter ", k, "max= ",maxval(u_new)
    temp = u
    u = u_new
    u_new = u
  enddo 

end program

subroutine zero(array)
  real, dimension(2048,2048)::array
  do i=1, 2048
    do j=1, 2048
      array(i,j)=0.0
    enddo
  enddo

end subroutine zero

subroutine random(array)
  real, dimension(2048,2048)::array
  do i=2,2047
    do j=2, 2047
      array(i,j) = rand()
    enddo
  enddo
end subroutine random

subroutine maximum(array)
  real, dimension(2048,2048)::array
  real :: maxi = 0.0
  do i=1, 2048
    do j=1, 2048
      if(array(i,j).gt.maxi) then
        maxi=array(i,j)
      end if
    end do
  end do
  write (*,*) "max=", maxi
end subroutine maximum

File: comparepythoncfortran/jacobiRelax/jacobirelax.c
/*
 * =====================================================================================
 *
 *       Filename:  jacobirelax.c
 *
 *    Description:  
 *
 *        Version:  1.0
 *        Created:  17/04/14 19:52:03
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  YOUR NAME (), 
 *   Organization:  
 *
 * =====================================================================================
 */
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#define SIZE 2048

double max(double * array){
  int i;
  double max=0.0;
  for(i=0; i< SIZE*SIZE; i++)
    if(array[i]>max)
      max=array[i];
  return max;
}

void initializeCero(double * array){
  int i;
  for(i=0; i<SIZE*SIZE; i++)
    array[i]=0.0;
}

void interiorRandom(double * array){
  int i,j;
  for(i=1; i<SIZE-1;i++)
    for(j=1; j<SIZE-1; j++)
      array[i*SIZE+j]=(float)rand()/(float)RAND_MAX;
}

void printMatrix(double * array){
  int i,j;
  for(i=1; i<SIZE-1; i++){
    for(j=1; j<SIZE-1; j++)
      printf("%f\t",array[i*SIZE+j]);
  }
}

int main(){
  double h2 = 1.0/SIZE;
  double  * u;
  double  * u_new;
  double  * temp;
  double  * f;
  int i,j,k;
  u     = (double*) malloc(SIZE*SIZE*sizeof(double));
  u_new = (double*) malloc(SIZE*SIZE*sizeof(double));
  f     = (double*) malloc(SIZE*SIZE*sizeof(double));

  /* Inizializamos  a cero */
  initializeCero(u);
  initializeCero(f);
  initializeCero(u_new);

  /* Inicializamos random en el interior de u */
  interiorRandom(u);
  printf("Initial=%f\n",max(u));

  for(k=0; k<10; k++){
    for(i=1; i < SIZE-1; i++){
      for(j=1; j< SIZE-1; j++){
        u_new[i*SIZE+j] = 0.25*(h2*f[i*SIZE+j] + u[(i-1)*SIZE+j] + u[(i+1)*SIZE+j] + 
                                     + u[i*SIZE+j-1]   + u[i*SIZE+j+1]); 
      }
    }
    temp=u;
    u=u_new;
    u_new=temp;
    printf("Iter %d: max=%f\n",k,max(u_new));
  }
}

